File: src/types/knowledge-domain/graph.ts
export interface KnowledgeGraphBuildInput {
  projectId: number;
  projectVersionId: number;
  /** 默认只投影已确认关系；开启后会额外显示待确认候选。 */
  includeUnconfirmed?: boolean;
}

export interface KnowledgeGraphBuildResult {
  buildId: number;
  buildKey: string;
  projectId: number;
  projectVersionId: number;
  nodeCount: number;
  edgeCount: number;
  reused: boolean;
}

export interface KnowledgeGraphQueryInput {
  projectId: number;
  projectVersionId: number;
  /** 留空时返回当前范围的有界总览。 */
  rootEntityKey?: string | null;
  /** 与实体标识共同确定根节点，避免同名异类型实体被合并。 */
  rootEntityType?: string | null;
  /** 省略时由后端限制为一层。 */
  depth?: number;
  /** 省略时由后端限制为 100 个节点。 */
  nodeLimit?: number;
  includeUnconfirmed?: boolean;
}

export interface KnowledgeGraphNode {
  id: number;
  entityType: string;
  entityKey: string;
  label: string;
}

export interface KnowledgeGraphEdge {
  id: number;
  fromNodeId: number;
  relationType: string;
  toNodeId: number;
  evidence: Record<string, unknown>;
  confidence: number;
  confirmed: boolean;
  sourceRelationRef: string;
}

export interface KnowledgeGraphProjection {
  buildId: number;
  buildKey: string;
  projectId: number;
  projectVersionId: number;
  nodes: KnowledgeGraphNode[];
  edges: KnowledgeGraphEdge[];
  truncated: boolean;
}

/** 旧关系 Command 的兼容载荷；图谱投影完成前仍需保留来源关系的事实字段。 */
export interface UpsertKnowledgeGraphRelationInput {
  id?: number | null;
  projectId?: number | null;
  releaseId?: number | null;
  documentVersionId?: number | null;
  snapshotId?: number | null;
  sensitivity: string;
  fromType: string;
  fromKey: string;
  relationType: string;
  toType: string;
  toKey: string;
  evidence: Record<string, unknown>;
  confidence: number;
  confirmed: boolean;
  source: string;
}

export interface ListKnowledgeGraphRelationsInput {
  entityType?: string | null;
  entityKey?: string | null;
  projectIds?: number[];
  releaseIds?: number[];
  sensitivities?: string[];
  confirmedOnly?: boolean | null;
  limit?: number | null;
}

File: src-tauri/src/models/knowledge_domain/graph.rs
use serde::{Deserialize, Serialize};

/// 图谱构建始终锁定到一个已存在的项目版本；知识来源只来自本地已入库文档和关系，
/// 不会在构建过程中把正文发送给远程服务。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KnowledgeGraphBuildInput {
    pub project_id: i64,
    pub project_version_id: i64,
    #[serde(default)]
    pub include_unconfirmed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KnowledgeGraphBuildResult {
    pub build_id: i64,
    pub build_key: String,
    pub project_id: i64,
    pub project_version_id: i64,
    pub node_count: u32,
    pub edge_count: u32,
    /// 相同来源哈希已是当前启用投影时直接复用，避免构建过程中短暂替换可见图谱。
    pub reused: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KnowledgeGraphQueryInput {
    pub project_id: i64,
    pub project_version_id: i64,
    #[serde(default)]
    pub root_entity_key: Option<String>,
    #[serde(default)]
    pub root_entity_type: Option<String>,
    #[serde(default = "default_depth")]
    pub depth: u8,
    #[serde(default = "default_node_limit")]
    pub node_limit: u32,
    #[serde(default)]
    pub include_unconfirmed: bool,
}

fn default_depth() -> u8 {
    1
}

fn default_node_limit() -> u32 {
    100
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KnowledgeGraphNode {
    pub id: i64,
    pub entity_type: String,
    pub entity_key: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KnowledgeGraphEdge {
    pub id: i64,
    pub from_node_id: i64,
    pub relation_type: String,
    pub to_node_id: i64,
    pub evidence: serde_json::Value,
    pub confidence: f64,
    pub confirmed: bool,
    pub source_relation_ref: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KnowledgeGraphProjection {
    pub build_id: i64,
    pub build_key: String,
    pub project_id: i64,
    pub project_version_id: i64,
    pub nodes: Vec<KnowledgeGraphNode>,
    pub edges: Vec<KnowledgeGraphEdge>,
    pub truncated: bool,
}
